### PR TITLE
Fix:The SplitContainer control cannot be deleted successfully under ToolStripContainer tab #11793

### LINKED ISSUE #11793: The SplitContainer control cannot be deleted successfully under ToolStripContainer tab
### .NET version

feature/10.0 branch of Winforms repo

### Did it work in .NET Framework?

Not tested/verified

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No, this was changed after the PR https://github.com/dotnet/winforms/pull/11678

### Issue description

**Actual result in [[DemoConsole](https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/DesignSurface/DemoConsole)] test app:**
The SplitContainer control cannot be deleted successfully under ToolStripContainer tab 
image
GHIssue4

**Expected result:**
The SplitContainer control can be deleted successfully under ToolStripContainer tab 

### Steps to reproduce

Test sample: [DemoConsole](https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms/tests/IntegrationTests/DesignSurface/DemoConsole) bases on feature/10.0 branch in Winforms repo.

### PR DESCRIPTION
**Pull request by Epica3055**
Fixes #11793

## Proposed changes

When `SplitContainer` is deleted, two things will happen
1.`SplitContainerDesigner.Dispose(bool)` will be invoked,and eventually `ComponentDesigner.Dispose(bool)` will be invoked.
So in the end, `SplitContainerDesigner.Component` will be null

https://github.com/dotnet/winforms/blob/ab9eb71f903d7771849887a15ab5ff3f37ba4c00/src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs#L452-L464

2. `SplitContainerDesigner.ActionList` will be accessed.

https://github.com/dotnet/winforms/blob/ab9eb71f903d7771849887a15ab5ff3f37ba4c00/src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/SplitContainerDesigner.cs#L39-L42

https://github.com/dotnet/winforms/blob/ab9eb71f903d7771849887a15ab5ff3f37ba4c00/src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/SplitContainerDesigner.OrientationActionList.cs#L20-L35

please see line 21,`SplitContainerDesigner.Component` is accessed.

https://github.com/dotnet/winforms/blob/ab9eb71f903d7771849887a15ab5ff3f37ba4c00/src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs#L172

when `Component` is accessed and it is null, exception will be thrown.

<hr>
InProcDesigner works fine because:

```c#
class ComponentDesigner
        protected virtual void Dispose(bool disposing) {
            if (disposing) {
                IComponentChangeService cs = (IComponentChangeService)GetService(typeof(IComponentChangeService));
                if (cs != null)
                {
                    cs.ComponentRename -= new ComponentRenameEventHandler(this.OnComponentRename);
                }
            
                component = null;
                inheritedProps = null;
            }
        }


        public IComponent Component {
            get {
                return component;
            }
        }
```

OOP Designer works fine because:

```c#
class ComponentDesigner
    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            ComponentChangeService.ComponentRename -= OnComponentRename;
        }

        base.Dispose(disposing);
    }


    public IComponent Component
    {
        get => _component ?? throw new InvalidOperationException(SR.DesignerUninitialized);
        private set => _component = value.OrThrowIfArgumentIsNull();
    }
```

### Before

image
GHIssue4

### After

issue_11793

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/winforms/pull/11830)

### CODE REVIEW

**Comment by Tanya-Solyanik** [comment]
>SplitContainerDesigner.ActionList will be accessed.

Why is ActionList accessed? 
I think that InProc designer works fine because component is not null, not because there is no explicit exception.
OOP designer sets component to null in the base.Dispose() call

```
    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            ComponentChangeService.ComponentRename -= OnComponentRename;
        }

        base.Dispose(disposing);
    }
```

**Comment by Epica3055** [comment]
@Tanya-Solyanik 

> Why is ActionList accessed?

`DesignerActionService` has a `OnComponentRemoved` event handler, in it's handler and the following logic, `ActionList` will be accessed.

https://github.com/microsoft/winforms-designer/blob/ba05432142bcb4472e909965f46b555d1bf6ef38/src/Microsoft.DotNet.DesignTools.Server/Designers/Actions/DesignerActionService.cs#L304-L322


https://github.com/microsoft/winforms-designer/blob/ba05432142bcb4472e909965f46b555d1bf6ef38/src/Microsoft.DotNet.DesignTools.Server/Designers/Actions/DesignerActionService.cs#L79-L81

> I think that InProc designer works fine because component is not null, not because there is no explicit exception.

Exactly.



> OOP designer sets component to null in the base.Dispose() call

image

image

**Comment by Tanya-Solyanik** [comment]
InProc designer checks if [Component is null or not before creating the action list](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=/src/vsip/Designer/WinForms/System/WinForms/Design/SplitContainerDesigner.cs&version=GBmain&line=615&lineEnd=631&lineStartColumn=1&lineEndColumn=14&lineStyle=plain&_a=contents). In this designer we implemented nullability incorrectly, there is no way to verify if the designer had been disposed before creating the action list. Your fix forces action list creation when it's not needed (the designer is gone, action list can't be displayed)


Exception call stack for reference:

```
System.Windows.Forms.Design.dll!System.Windows.Forms.Design.SplitContainerDesigner.OrientationActionList.OrientationActionList(System.Windows.Forms.Design.SplitContainerDesigner owner) Line 21	C#
System.Windows.Forms.Design.dll!System.Windows.Forms.Design.SplitContainerDesigner.ActionLists.get() Line 39	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.ComponentDesigner.CDDesignerCommandSet.GetCommands(string name) Line 22	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerCommandSet.ActionLists.get() Line 17	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerActionService.GetComponentDesignerActions(System.ComponentModel.IComponent component, System.ComponentModel.Design.DesignerActionListCollection actionLists) Line 171	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerActionService.GetComponentActions(System.ComponentModel.IComponent component, System.Windows.Forms.Design.ComponentActionsType type) Line 149	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerActionService.GetComponentActions(System.ComponentModel.IComponent component) Line 138	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerActionService.Remove(System.ComponentModel.IComponent comp) Line 299	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerActionService.OnComponentRemoved(object source, System.ComponentModel.Design.ComponentEventArgs componentEventArgs) Line 279	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerHost.RemoveFromContainerPostProcess(System.ComponentModel.IComponent component) Line 681	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerHost.Remove(System.ComponentModel.IComponent component) Line 614	C#
System.Windows.Forms.Design.dll!System.ComponentModel.Design.DesignerHost.System.ComponentModel.Design.IDesignerHost.DestroyComponent(System.ComponentModel.IComponent component) Line 1009	C#
System.Windows.Forms.Design.dll!System.Windows.Forms.Design.CommandSet.OnMenuDelete(object sender, System.EventArgs e) Line 1753	C#

```

I suggest changing nullability [here ](https://github.com/dotnet/winforms/blob/7006c1c2c5515bc4b648e5b3c2ea6a604867e0b9/src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs#L172)

    public IComponent? Component => _component;

And testing for null before creating the action list.

**Comment by Tanya-Solyanik** [comment]
@Epica3055 - this seems like too many null-forgiving operators. Maybe add a method that tests if ComponentDesigner._component is null instead.

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good, please add a unit test for this scenario.

**Comment by Tanya-Solyanik** [comment]
@Epica3055  - this error was introduced in NET9 release, we should fix it in main, please retarget this PR.

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good, but could you please add a unit test?

**Review by LeafShi1 — COMMENTED** [review]
LGTM!

File: src/System.Windows.Forms.Design/tests/UnitTests/SplitContainerDesignerTests.cs [hunk 1]
```diff
@@ -39,4 +45,22 @@ public void SplitContainerDesigner_AssociatedComponentsTest()
...
+
+        var mockSite = GetMockSize(splitContainer, splitContainerDesigner);
+        splitContainer.Site = mockSite.Object;
+
+        splitContainerDesigner.Initialize(splitContainer);
+
+        splitContainerDesigner.Dispose();
+
+        Action action = () => _ = splitContainerDesigner.ActionLists;
+        action.Should().NotThrow();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/tests/UnitTests/SplitContainerDesignerTests.cs:64]: Could you please make verification of the ActionLists property more specific? For example
`splitContainerDesigner.ActionLists.Should().BeEmpty()`

**Comment by Tanya-Solyanik** [comment]
> @Epica3055 - this error was introduced in NET9 release, we should fix it in main, please retarget this PR.

I double-checked, this regression was introduced when porting this designer, the fix does not qualify for RC1, I'm retargeting branch to feature/10

**Review by JeremyKuhne — APPROVED** [review]
:shipit: Modulo @Tanya-Solyanik's test comments.

File: src/System.Windows.Forms.Design/tests/UnitTests/SplitContainerDesignerTests.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Design/tests/UnitTests/SplitContainerDesignerTests.cs:64]: Now you can delete lines 63 and 64, if `ActionLists` throws, then we can't get the count and line 65 will fail.

**Review by Tanya-Solyanik — APPROVED** [review]
Thank you!

**Pull request merged**